### ISSUE TITLE
Investigate precision difference in tutorial `03-matrix-multiplication.py` and `08-experimental-block-pointer.py`

### ISSUE THREAD

**Issue by etiotto** [intel/intel-xpu-backend-for-triton#468]
The tutorial `03-matrix-multiplication.py` code was modified downstream to increase the tolerance used to check that the matmul kernel compiled by Triton yields the same results as the PyTorch implementation `torch.matmul`.  Investigate whether the Triton result is correct.

**Comment by etiotto** [comment]
The reason for the difference in results between the Triton implementation and Torch implementation is due to the algorithmic differences which lead to floating point errors being accumulated in different ways. 

The `tl.dot` operation in the Triton kernel uses `float32` precision when the input matrices use `float16` precision (unless the `out_dtype=tl.float32` is explicitly used). However the result of the following loop is then converted to `float16`:

```
    accumulator = tl.zeros((BLOCK_SIZE_M, BLOCK_SIZE_N), dtype=tl.float32)
    for k in range(0, tl.cdiv(K, BLOCK_SIZE_K)):
        # Load the next block of A and B, generate a mask by checking the K dimension.
        # If it is out of bounds, set it to 0.
        a = tl.load(a_ptrs, mask=offs_k[None, :] < K - k * BLOCK_SIZE_K, other=0.0)
        b = tl.load(b_ptrs, mask=offs_k[:, None] < K - k * BLOCK_SIZE_K, other=0.0)
        # We accumulate along the K dimension.
        accumulator += tl.dot(a, b)   <<<<<< uses float32 even when a,b are float16
        # Advance the ptrs to the next K block.
        a_ptrs += BLOCK_SIZE_K * stride_ak
        b_ptrs += BLOCK_SIZE_K * stride_bk
    # You can fuse arbitrary activation functions here
    # while the accumulator is still in FP32!
    if ACTIVATION == "leaky_relu":
        accumulator = leaky_relu(accumulator)
    c = accumulator.to(tl.float16)    <<<<<<  float32 results converted to float16
```

In order to evaluate the quality of the Triton result we have first changed the tutorial to use input matrices with `float32` element type. This entails removing the conversion to float16 in the kernel (`accumulator.to(tl.float16)`).  The Triton result then matches the pytorch (`torch.matmul` result). Furthermore it also matches the result computed via a sequential computation (using a loop nest). 

We then changed the tutorial to use `float16` precision in the Triton kernel (by adding out_dtype=tl.float16 to the tl.dot operation). With that change the Triton result matches the result obtained via sequential computation (loop nest) but, it did not match the results from `torch.matmul`. 

This experiments prove that the Triton implementation is correct when the computation is carried out fully using `float32` and/or `float16` precision.

**Comment by ienkovich** [comment]
I also experimented with manual fp32 computation for one input row/column for matmul with the following conversion to fp16. I did the computation in direct and reverse orders and fp32 and the results were `[-39.32813262939453, -39.32810592651367]`. Here, the difference is small enough and meets the required tolerance level. But if we convert these results to fp16 we get `[-39.34375, -39.3125]`. Thus, the difference in the fifth digit for the fp32 result transforms into `0.03125` difference for fp16 results exceeding the originally used tolerance of `0.01`. We would inevitably encounter such problems in other tutorials leading to mismatch. I believe we should never try to use `rtol=0` for the comparison. I don't know yet what the appropriate value for fp16 is, will do some research on this.
